### ISSUE TITLE
feat: thinking-aware constrained decoding (lm-format-enforcer + reasoning coexistence)

### ISSUE THREAD

**Issue by Thump604** [waybarrios/vllm-mlx#378]
## Problem

Constrained decoding via lm-format-enforcer (#362) and thinking/reasoning mode were mutually exclusive. When both active, the enforcer's FSM saw `<think>` as the first token, rejected it (not valid JSON), got stuck, and disabled itself. The server worked around this by forcing `enable_thinking=false` when constrained decoding was active.

### lm-format-enforcer path: RESOLVED

Three merged PRs fix the lm-format-enforcer + thinking coexistence:

- **#436** — `ThinkingAwareLogitsProcessor` with 4-phase state machine (IDLE/THINKING/TRANSITIONING/CONTENT), `thinking_token_budget` enforcement, and inner processor delegation
- **#458** — Root cause fix for token duplication in `_step()` that caused the enforcer to crash, plus `_ThinkingAwareLogitsProcessor` wrapper for JSON-phase gating after `</think>`
- **#472** — Skip preamble text before activating JSON schema enforcer, with `_json_scan_offset` tracking so `{` inside reasoning is ignored

### Outlines path: BLOCKED on two upstream issues

1. **[dottxt-ai/outlines#1847](https://github.com/dottxt-ai/outlines/issues/1847)** — `TransformerTokenizer.__init__` reads `eos_token_id`, `eos_token`, `all_special_tokens` from the raw `tokenizers.Tokenizer` backend, but modern `transformers` (tokenizers 0.25+) only exposes these on `PreTrainedTokenizer`. Outlines maintainer (RobinPicard) has opened a fix PR.

2. **[ml-explore/mlx-lm#1175](https://github.com/ml-explore/mlx-lm/issues/1175)** — MTP `_step_mtp` calls `_process_and_sample(prev_tokens, mtp_logits)` without the backbone's just-sampled token in `prev_tokens`. Stateful processors (Outlines Guide FSM) advance by the wrong token and enter invalid state. Blocks Outlines structured output when MTP is enabled.

## Acceptance criteria

This issue is done when:
- [x] lm-format-enforcer + thinking mode coexist without forcing `enable_thinking=false`
- [ ] Outlines + thinking mode coexist (blocked on dottxt-ai/outlines#1847)
- [ ] Outlines + MTP coexist (blocked on ml-explore/mlx-lm#1175)

**Comment by janhilgard** [comment]
This is resolved by two merged PRs:

- **#436** — `ThinkingAwareLogitsProcessor` with 4-phase state machine (IDLE/THINKING/TRANSITIONING/CONTENT), `thinking_token_budget` enforcement, and inner processor delegation (`constrained/thinking_processor.py`)
- **#458** — Root cause fix for token duplication in `_step()` that caused the enforcer to crash, plus `_ThinkingAwareLogitsProcessor` wrapper in `server.py` for JSON-phase gating after `</think>`

Both are merged and coexist in `main`. Thinking + constrained decoding now work together without forcing `enable_thinking=False`.

**Comment by Thump604** [comment]
Resolved by three merged PRs:
- #436: ThinkingAwareLogitsProcessor with 4-phase state machine and thinking_token_budget enforcement
- #458: Token duplication fix in _step() that caused enforcer crash, plus _ThinkingAwareLogitsProcessor wrapper for JSON-phase gating after </think>
- #472: Skip preamble text before activating JSON schema enforcer, with _json_scan_offset tracking

Thinking + constrained decoding now coexist without forcing enable_thinking=False.

**Comment by Thump604** [comment]
Note: two upstream blockers remain for full Outlines structured output support with MTP:

1. **dottxt-ai/outlines#1847** -- `TransformerTokenizer` reads attributes from raw tokenizers backend that modern `transformers` doesn't set. Outlines maintainer has a fix PR open.
2. **ml-explore/mlx-lm#1175** -- MTP `_step_mtp` passes stale `prev_tokens` to logits processors, causing Outlines Guide FSM to advance by the wrong token.

These don't block the thinking+constrained decoding work that closed this issue (#436, #458, #472 via lm-format-enforcer), but they block the Outlines path specifically when MTP is enabled.

**Comment by Thump604** [comment]
Reopening. The lm-format-enforcer path is resolved (#436, #458, #472), but the Outlines structured output path is still blocked by two upstream issues:

1. dottxt-ai/outlines#1847 -- TransformerTokenizer attribute error on modern transformers (fix PR open upstream)
2. ml-explore/mlx-lm#1175 -- MTP stale prev_tokens breaks stateful logits processors like Outlines Guide FSM

Keeping this open until both Outlines and lm-format-enforcer paths work with thinking + MTP enabled.

**Comment by Thump604** [comment]
Issue #378 status update:

- dottxt-ai/outlines#1847 is no longer an active item; it was closed by dottxt-ai/outlines#1848 on 2026-05-04.
- The remaining open issue for the Outlines + MTP path is ml-explore/mlx-lm#1175; ml-explore/mlx-lm#990 contains proposed MTP work but is still open as of 2026-05-16.
- Runtime also has local Qwen 3.6 template tests covering the unclosed <think> before <tool_call> case; that is separate from closing this vllm-mlx issue.

I would keep #378 open for the Outlines + MTP / stateful logits-processor path, and not treat dottxt-ai/outlines#1847 as a current blocker anymore. Not claiming full Outlines structured-output support or full #378 closure.

**Comment by janhilgard** [comment]
Status check on the blockers, since this has been quiet for a while.

| dependency | state |
|---|---|
| dottxt-ai/outlines#1847 | closed by outlines#1848, **merged** |
| ml-explore/mlx-lm#1175 | **open**, last touched 2026-04-28 |
| ml-explore/mlx-lm#990 | **open**, but active — updated 2026-08-24 |

So one blocker is gone and one has been static for four months.

## What vllm-mlx does today

I checked rather than assumed: the MLLM path already refuses to draft when request-local logits processors are present.

```python
logits_processors_bypass = logits_processors is not None and any(logits_processors)
if prefill_bypass or no_active_batch_bypass or logits_processors_bypass:
    ...
    _bypass_counts["logits_processors"] += 1
```

That is the correct response to mlx-lm#1175 — the issue is that `mtp_generate_step` hands processors a `prev_tokens` that lacks the backbone's just-sampled token, so a stateful FSM advances on the wrong token and enters an invalid state. Not drafting at all cannot hit that. Constrained requests therefore get correct output today, with zero speculative acceleration, and the bypass counter makes it observable rather than silent.

## Why I think this de-risks the issue rather than blocking it

The thing that changes the calculus: on hybrid linear-attention models, MTP is not actually buying throughput, so the acceleration the bypass gives up is close to nothing.

Measured on Qwen3.8-27B (M3 Ultra, native MTP head, paired A/B alternating per prompt so machine load hits both arms equally): **0.95–0.96x** with MTP against a matched non-MTP baseline. Raising draft depth makes it worse — k=2 costs +38 % per step for +21 % tokens, k=3 costs +71 % for +31 % — because the gated-delta recurrent scan is per-token work rather than one batched op, so the verify forward scales with k instead of amortising. Step profile at k=1: verify forward 69 %, reject/replay 19 %, draft 10 %, state snapshot 1.8 %. Full write-up in #710.

If that holds for the models this issue cares about, then "constrained decoding does not get MTP" costs a few percent, not a speedup. mlx-lm#1175 stops being a prerequisite and becomes an optimisation that can land whenever upstream gets to it.

I would suggest decoupling: land the constrained-decoding work against the existing bypass, and track mlx-lm#1175 separately as the thing that would let constrained requests draft too. That unblocks four months of waiting on an upstream issue nobody is moving.

Caveat on scope: my numbers are from one model family on one machine. If the constrained-decoding use case here is aimed at pure-attention models, the verify forward amortises properly there and MTP may well be a real win — in which case #1175 stays load-bearing and this reasoning does not apply. Happy to measure a specific configuration if that would settle it.

**Comment by Thump604** [comment]
Fresh source/dependency check:

- `dottxt-ai/outlines#1847` is fixed by merged `#1848` and included in Outlines 1.2.13.
- Current vllm-mlx does not ship an Outlines backend; it was removed in `dc421c6`/`d0a8d2a`. Restoring it is therefore an implementation and qualification task, not a dependency bump.
- The current MLLM path correctly bypasses MTP when request-local logits processors are present, preserving constrained-decoding correctness without speculative acceleration.
- `ml-explore/mlx-lm#1175` and proposed PR `#990` remain open.

Keeping this issue open under its existing acceptance criteria. Closure still needs a current Outlines backend plus live Outlines+thinking and Outlines+MTP qualification, or a separate maintainer decision to revise that scope.

**Comment by janhilgard** [comment]
Assignee follow-up. @Thump604's 29 Aug check is right on every point, and I think it leads to a conclusion worth stating plainly: **this issue's remaining criteria are no longer about thinking-aware constrained decoding at all.**

Fresh state of the three dependencies:

| dependency | state |
|---|---|
| `dottxt-ai/outlines#1847` | closed by merged `#1848`, shipped in Outlines 1.2.13 |
| `ml-explore/mlx-lm#1175` | **open**, untouched since 2026-04-28 — four months static |
| `ml-explore/mlx-lm#990` | **open** but genuinely active, updated 2026-09-01 |

And the local state, checked rather than assumed:

```
vllm_mlx/mllm_batch_generator.py:2340  logits_processors_bypass = logits_processors is not None and any(...)
vllm_mlx/mllm_batch_generator.py:2364  if logits_processors_bypass:
vllm_mlx/mllm_batch_generator.py:2365      _bypass_counts["logits_processors"] += 1
```

The MTP path still declines to draft whenever request-local logits processors are present, so constrained decoding stays correct with MTP enabled — it just does not get speculative acceleration. That is the right trade and it means mlx-lm#1175 is not currently costing correctness, only speed.

## Why I think the scope should be revised

The issue title and problem statement are about `<think>` and constrained decoding being mutually exclusive. That is **solved** — #436, #458 and #472 landed the four-phase `ThinkingAwareLogitsProcessor`, the `_step()` token-duplication root cause, and the preamble-skip with `_json_scan_offset`, and the server no longer forces `enable_thinking=false`. The first acceptance box is ticked and has been for months.

The two open boxes both say "Outlines". But vllm-mlx has not shipped an Outlines backend since `dc421c6`/`d0a8d2a`, and #753 — which I approved today — makes that explicit at the API boundary by returning 422 for any request that names `outlines` as its backend. So those boxes cannot be ticked by fixing a bug; they need a backend to be written, qualified against thinking, and qualified against MTP. That is a feature, not the tail of this one.

Keeping them in the same issue has a practical cost: #378 reads as "thinking-aware constrained decoding is unfinished", when the thing it was filed about works and ships. Anyone triaging by title gets the wrong picture, and the real remaining work — "restore an Outlines constrained-decoding backend" — has no issue of its own and no visible demand attached to it.

## Proposal

Close #378 as done against its stated problem, and open a separate issue for the Outlines backend, referencing #753 for the current API behaviour and mlx-lm#1175 as a known limitation for its MTP story. If nobody files that issue, that is itself a useful signal about whether the backend is wanted back.

Alternative, if you would rather not close: strike the two Outlines boxes from the acceptance criteria and retitle to match what is left.

@waybarrios either works for me and it is a maintainer call, so I am not closing it myself. Say which and I will do the bookkeeping.

One thing I would not do is keep waiting on mlx-lm#1175. Four months static, and the local bypass means we are not blocked on it for correctness.

**Comment by Thump604** [comment]
Thanks, Jan. I agree that restoring Outlines is a separate backend feature, but I need to correct my earlier status before we close this.

On current main `f2d3ad3`, [`_attach_response_format_logits_processor()`](https://github.com/waybarrios/vllm-mlx/blob/f2d3ad394eb4f2b5e0b1352364eddfed41e390e1/vllm_mlx/server.py#L723-L740) explicitly sets `enable_thinking=False`, including the template flag when present. Both Chat Completions and Anthropic preparation call it when a response-format processor is attached. [`test_response_format_attachment_suppresses_thinking`](https://github.com/waybarrios/vllm-mlx/blob/f2d3ad394eb4f2b5e0b1352364eddfed41e390e1/tests/test_thinking_aware_processor.py#L124-L146) asserts this behavior.

That came from my merged #539 (`ced8cb1a`), which kept JSON constraints in direct control after the thinking wrapper allowed direct JSON to bypass them. My August status carried forward the earlier #436/#458/#472 resolution without accounting for that later change. Sorry for the misleading status. The first acceptance box is not currently satisfied for these server paths.

Two other distinctions matter: #753 is approved by you but still open, so its explicit Outlines rejection is not yet on main; and the MLLM logits-processor bypass avoids that speculative path, but does not prove thinking coexistence or successful Outlines decoding.

I am correcting the issue body and keeping it open. The next local task is to reconcile the current `response_format`/thinking contract with #539's safety requirement, using the real preparation paths and preserving direct-JSON constraints and streaming content handling. It is not to simply remove the flags or wait on mlx-lm#1175. Outlines can be moved to a linked feature issue with maintainer agreement; no backend should be built merely to tick a stale box. This is source/test-definition inspection, not a new live qualification.
